core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java

```diff
@@ -137,8 +139,53 @@ public CalciteCatalogReader withSchemaPath(List<String> schemaPath) {
     return config;
   }
 
+  /**
+   * Gets the macro with the given name if it exists.
+   *
+   * @param names The name of the macro, list represents possibility of a
+   *              compound identifier
+   *
+   * @return The found macro, or null
+   */
+  public Macro getMacro(List<String> names) {
```

Comment:
This looks like it has a lot of duplicate code to `getFunctionsFrom()`. Is there a way refactor shared logic out into a common helper function that they can both call?